### PR TITLE
add ray cluster mode in init_orca_context

### PR DESCRIPTION
**Pull request by yangw1234**
Usage:

Init Ray Driver:
```python
init_orca_context("ray", address="xxxx:xxx")
```

Init Ray Client:
```python
init_orca_context("ray", address="ray://xxxx:xxx")
```

Limitations:
1. Only support data_creator

### CODE REVIEW

File: python/orca/src/bigdl/orca/ray/ray_on_spark_context.py [hunk 1]
```diff
@@ -0,0 +1,625 @@
+#
```
Comment (yangw1234) [inline python/orca/src/bigdl/orca/ray/ray_on_spark_context.py:1]: This file just copied the origin raycontext.py

File: python/orca/src/bigdl/orca/common.py [hunk 2]
```diff
@@ -197,102 +197,110 @@ def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
...
-    if activate_sc:
-        if cluster_mode is not None and cluster_mode != "spark-submit":
-            warnings.warn("Use an existing SparkContext, " +
-                          "cluster_mode is determined by the existing SparkContext", Warning)
-        from bigdl.dllib.nncontext import init_nncontext
-        sc = init_nncontext(conf=None, spark_log_level="WARN", redirect_spark_log=True)
+    if cluster_mode == "ray":
+        import ray
+        from bigdl.orca.ray import RayContext
+        ray_ctx = RayContext(cluster_mode="ray", **kwargs)
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:203]: pass the parameters of `cores`, `num_nodes` (and maybe `memory`) to RayContext

File: python/orca/src/bigdl/orca/ray/raycontext.py [hunk 3]
```diff
@@ -14,612 +14,46 @@
...
-            return int(local_symbol)
+    def __init__(self, cluster_mode="ray_on_spark", **kwargs):
+
+        self.cluster_mode = cluster_mode
+
+        if cluster_mode == "ray_on_spark":
+            from bigdl.orca.ray.ray_on_spark_context import RayOnSparkContext
+            self._ray_on_spark_context = RayOnSparkContext(**kwargs)
+            self.num_ray_nodes = self._ray_on_spark_context.num_ray_nodes
+            self.ray_node_cpu_cores = self._ray_on_spark_context.ray_node_cpu_cores
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/ray/raycontext.py:28]: use the `cores` and `num_nodes` parameters in `init_orca_context` instead.

**Comment by jason-dai** [comment]
Maybe we should not overload `cluster_model` for this. Instead, add a `runtime` argument:
1) `cluster_model` = "local", "k8s", "yarn", etc., and default to None
2) `runtime` = "spark", "ray", "spark_and_ray", etc., and default to "spark"

We can then have something like:
```python
def init_orca_context(cluster_mode=None, runtime="spark", ...):
    if runtime == "spark":
       ...
    elif runtime == "ray":
       assert cluster_mode == None, "..."
       ...
    elif
       raise ValueError(...)
```

File: python/orca/src/bigdl/orca/common.py [hunk 4]
```diff
@@ -197,102 +197,114 @@ def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
...
-            warnings.warn("Use an existing SparkContext, " +
-                          "cluster_mode is determined by the existing SparkContext", Warning)
-        from bigdl.dllib.nncontext import init_nncontext
-        sc = init_nncontext(conf=None, spark_log_level="WARN", redirect_spark_log=True)
+    if cluster_mode == "ray":
+        import ray
+        from bigdl.orca.ray import RayContext
+        ray_ctx = RayContext(cluster_mode="ray", cores=cores, num_nodes=num_nodes,
+                             **kwargs)
+        assert "address" in kwargs, "address must be specified if cluster_mode is ray"
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:205]: "ray_address" instead of "address"

File: python/orca/src/bigdl/orca/common.py [hunk 5]
```diff
@@ -197,102 +197,114 @@ def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
...
+        from bigdl.orca.ray import RayContext
+        ray_ctx = RayContext(cluster_mode="ray_on_spark", cores=cores, num_nodes=num_nodes,
+                             sc=sc, **ray_args)
+        if init_ray_on_spark:
+            driver_cores = 0  # This is the default value.
+            ray_ctx.init(driver_cores=driver_cores)
+        return sc
 
 
 def stop_orca_context():
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:310]: Need to do anything for `stop_orca_context`?

File: python/orca/src/bigdl/orca/ray/raycontext.py [hunk 6]
```diff
@@ -15,436 +15,49 @@
...
-            # psutil may encounter AccessDenied or ZombieProcess exceptions
-            # when it's trying to visit some MacOS core services
-            if psutil.MACOS:
-                continue
-            else:
-                raise Exception("List process with list2cmdline failed!")
+    def __init__(self,
+                 cluster_mode="ray_on_spark",
+                 cores=None,
+                 num_nodes=None,
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/ray/raycontext.py:25]: `cores` and `num_nodes` should not be None

File: python/orca/src/bigdl/orca/common.py [hunk 7]
```diff
@@ -197,102 +197,114 @@ def init_orca_context(cluster_mode=None, cores=2, memory="2g", num_nodes=1,
...
-        cluster_mode = cluster_mode.lower()
-    activate_sc = SparkContext._active_spark_context is not None
-    if activate_sc:
-        if cluster_mode is not None and cluster_mode != "spark-submit":
-            warnings.warn("Use an existing SparkContext, " +
-                          "cluster_mode is determined by the existing SparkContext", Warning)
-        from bigdl.dllib.nncontext import init_nncontext
-        sc = init_nncontext(conf=None, spark_log_level="WARN", redirect_spark_log=True)
+    if cluster_mode == "ray":
+        import ray
```
Comment (jason-dai) [inline python/orca/src/bigdl/orca/common.py:201]: do you need to `import ray` here?

**Comment by yangw1234** [comment]
move to https://github.com/intel-analytics/BigDL/pull/3885

**Pull request closed without merging**